File: apps/worker/src/runLoop.ts
import type OpenAI from "openai";
import {
  type RunoffDb,
  type RunEvent,
  type BlueprintContent,
  BlueprintContentSchema,
  previousCompletedDocument,
  newId,
} from "@runoff/core";
import {
  executeRun,
  distillRun,
  type RunInteractions,
  type EngineIO,
  type RunInputMsg,
} from "@runoff/engine";
import { resolveRunSources } from "./resolveSources.js";

/** Atomically claim the oldest queued run (single statement; RETURNING gives us its identity). */
const CLAIM_SQL = `
UPDATE runs SET status='running', started_at=datetime('now')
WHERE id = (SELECT id FROM runs WHERE status='queued' ORDER BY created_at LIMIT 1)
RETURNING id, blueprint_id AS blueprintId, blueprint_rev AS blueprintRev, period`;

export function claimQueuedRun(
  db: RunoffDb,
): { id: string; blueprintId: string; blueprintRev: number; period: string | null } | undefined {
  const row = db.sqlite.prepare(CLAIM_SQL).get() as
    | { id: string; blueprintId: string; blueprintRev: number; period: string | null }
    | undefined;
  return row ?? undefined;
}

/**
 * The side-channel the engine drives for a single run. `emit` appends a run
 * event (seq = max+1) in one transaction and mirrors run status onto the
 * `runs` row (and, for flags, into the `flags` table) so the API can report
 * status without replaying events. `pollInputs` drains pending worker inputs.
 */
export function makeEngineIO(db: RunoffDb, runId: string): EngineIO {
  const nextSeq = db.sqlite.prepare("SELECT COALESCE(MAX(seq), 0) + 1 AS seq FROM run_events WHERE run_id = ?");
  const insertEvent = db.sqlite.prepare("INSERT INTO run_events (run_id, seq, type, payload) VALUES (?, ?, ?, ?)");
  const insertFlag = db.sqlite.prepare(
    "INSERT INTO flags (id, run_id, code, section_key, question, options) VALUES (?, ?, ?, ?, ?, ?)",
  );
  const setStatus = db.sqlite.prepare("UPDATE runs SET status=? WHERE id = ?");
  const finishComplete = db.sqlite.prepare(
    "UPDATE runs SET status='complete', finished_at=datetime('now'), stats=?, document=? WHERE id = ?",
  );
  const finishFailed = db.sqlite.prepare("UPDATE runs SET status='failed', finished_at=datetime('now') WHERE id = ?");

  const emitTx = db.sqlite.transaction((e: RunEvent) => {
    const { seq } = nextSeq.get(runId) as { seq: number };
    insertEvent.run(runId, seq, e.type, JSON.stringify(e));
    switch (e.type) {
      case "flag_raised":
        // The engine mints per-run flag ids (`flag_1`, `flag_2`, …); the `flags`
        // table is globally keyed, so namespace the row id by run to avoid a
        // cross-run primary-key collision. The event payload (and thus the UI's
        // in-flight projection) keeps the bare id; the API reads the row id back.
        insertFlag.run(`${runId}_${e.flagId}`, runId, e.code, e.sectionKey, e.question, JSON.stringify(e.options));
        break;
      case "paused":
        setStatus.run("paused", runId);
        break;
      case "resumed":
        setStatus.run("running", runId);
        break;
      case "run_completed":
        finishComplete.run(JSON.stringify(e.stats), JSON.stringify(e.document), runId);
        break;
      case "run_failed":
        finishFailed.run(runId);
        break;
    }
  });

  const selectInputs = db.sqlite.prepare(
    "SELECT id, kind, payload FROM run_inputs WHERE run_id = ? AND consumed_at IS NULL ORDER BY id",
  );
  const consumeInput = db.sqlite.prepare("UPDATE run_inputs SET consumed_at=datetime('now') WHERE id = ?");

  return {
    emit(e: RunEvent): void {
      // Immediate mode takes the write lock up front, so the MAX(seq) read and
      // the insert see one snapshot — avoids SQLITE_BUSY_SNAPSHOT (which
      // busy_timeout does not retry) when the web app writes concurrently.
      emitTx.immediate(e);
    },
    pollInputs(): RunInputMsg[] {
      const rows = selectInputs.all(runId) as { id: number; kind: string; payload: string }[];
      return rows.map((r) => {
        consumeInput.run(r.id);
        const payload = JSON.parse(r.payload) as { text?: string; questionId?: string };
        return { kind: r.kind as RunInputMsg["kind"], text: payload.text, questionId: payload.questionId };
      });
    },
    sleep(ms: number): Promise<void> {
      return new Promise((resolve) => setTimeout(resolve, ms));
    },
  };
}

/**
 * Boot recovery: any run left `running`/`paused` by a crashed worker is
 * unrecoverable (the engine's in-memory state is gone), so mark it failed and
 * append a `run_failed` event. Returns how many runs were recovered.
 */
export function failStaleRuns(db: RunoffDb): number {
  const stale = db.sqlite
    .prepare("SELECT id FROM runs WHERE status IN ('running', 'paused')")
    .all() as { id: string }[];
  for (const { id } of stale) {
    makeEngineIO(db, id).emit({ type: "run_failed", error: "worker restarted mid-run" });
  }
  return stale.length;
}

/**
 * Claim and execute one queued run. Loads the pinned blueprint revision content
 * and its bound source files, runs the engine (which persists success and, on
 * failure, emits `run_failed` before throwing), and guards pre-engine throws
 * (e.g. a malformed revision) by marking the run failed itself.
 * Returns whether a run was claimed.
 */
export async function processOne(db: RunoffDb, client: OpenAI): Promise<boolean> {
  const claimed = claimQueuedRun(db);
  if (!claimed) return false;

  const io = makeEngineIO(db, claimed.id);
  try {
    const revRow = db.sqlite
      .prepare("SELECT content FROM blueprint_revisions WHERE blueprint_id = ? AND rev = ?")
      .get(claimed.blueprintId, claimed.blueprintRev) as { content: string } | undefined;
    if (!revRow) throw new Error(`blueprint revision not found: ${claimed.blueprintId}@${claimed.blueprintRev}`);
    const content = BlueprintContentSchema.parse(JSON.parse(revRow.content));

    const { files, gaps } = resolveRunSources(db, claimed.blueprintId, claimed.period);

    const runRow = db.sqlite
      .prepare("SELECT created_at AS createdAt FROM runs WHERE id = ?")
      .get(claimed.id) as { createdAt: string };
    const previous = previousCompletedDocument(db.sqlite, claimed.blueprintId, {
      runId: claimed.id,
      createdAt: runRow.createdAt,
    });

    const memoryRows = db.sqlite
      .prepare(
        `SELECT id, body, scope FROM memories
         WHERE status='active' AND (blueprint_id = ? OR (scope='project' AND project_id = (SELECT project_id FROM blueprints WHERE id = ?)))
         ORDER BY rowid`,
      )
      .all(claimed.blueprintId, claimed.blueprintId) as { id: string; body: string; scope: "blueprint" | "project" }[];

    await executeRun({
      client,
      content,
      files,
      io,
      blueprintRev: claimed.blueprintRev,
      previousDocument: previous?.document,
      memories: memoryRows,
      period: claimed.period,
      gaps,
    });
    // Success is already persisted by the `run_completed` emit handler.
    await distillCompletedRun(db, client, claimed.id, claimed.blueprintId, content);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    console.error(`[worker] run ${claimed.id} failed: ${message}`);
    // If the engine threw it already emitted `run_failed` (which set status).
    // For pre-engine throws (revision missing / zod parse) nothing marked it yet.
    const row = db.sqlite.prepare("SELECT status FROM runs WHERE id = ?").get(claimed.id) as
      | { status: string }
      | undefined;
    if (row && row.status !== "failed") {
      io.emit({ type: "run_failed", error: message });
    }
  }
  return true;
}

/**
 * Post-run learning: turn this run's human interventions into standing
 * memories. Read-side is the run's own event log + resolved flags; any failure
 * is logged and swallowed — distillation must never affect run status.
 */
async function distillCompletedRun(
  db: RunoffDb,
  client: OpenAI,
  runId: string,
  blueprintId: string,
  content: BlueprintContent,
): Promise<void> {
  try {
    const events = db.sqlite
      .prepare(
        "SELECT type, payload FROM run_events WHERE run_id = ? AND type IN ('steer_received','question_raised','question_answered') ORDER BY seq",
      )
      .all(runId) as { type: string; payload: string }[];
    // `question_answered` carries only {questionId, answer}; the question text
    // lives in the preceding `question_raised` event, so map ids -> text first.
    const questionText = new Map<string, string>();
    for (const e of events) {
      if (e.type !== "question_raised") continue;
      const p = JSON.parse(e.payload);
      if (typeof p.questionId === "string" && typeof p.question === "string") questionText.set(p.questionId, p.question);
    }
    const interactions: RunInteractions = { steers: [], answers: [], flagResolutions: [] };
    for (const e of events) {
      const p = JSON.parse(e.payload);
      if (e.type === "steer_received" && typeof p.text === "string") interactions.steers.push(p.text);
      if (e.type === "question_answered" && typeof p.answer === "string") {
        interactions.answers.push({ question: questionText.get(p.questionId) ?? String(p.questionId ?? ""), answer: p.answer });
      }
    }
    const flagRows = db.sqlite
      .prepare("SELECT question, resolution FROM flags WHERE run_id = ? AND status = 'resolved' AND resolution IS NOT NULL")
      .all(runId) as { question: string; resolution: string }[];
    interactions.flagResolutions = flagRows.map((f) => ({ question: f.question, resolution: f.resolution }));

    if (!interactions.steers.length && !interactions.answers.length && !interactions.flagResolutions.length) return;

    // This blueprint's project, so project-scoped memories land on the right row.
    const projectId =
      (db.sqlite.prepare("SELECT project_id AS projectId FROM blueprints WHERE id = ?").get(blueprintId) as
        | { projectId: string }
        | undefined)?.projectId ?? "";

    // Dedup pool spans both scopes (dedup is on lowercased body, scope-agnostic).
    const existing = db.sqlite
      .prepare(
        `SELECT body, scope FROM memories WHERE blueprint_id = ? OR (scope='project' AND project_id = ?)`,
      )
      .all(blueprintId, projectId) as { body: string; scope: string }[];

    const fresh = await distillRun({
      client,
      title: content.title,
      sectionHeadings: content.sections.map((s) => s.heading),
      interactions,
      existing,
    });

    const insert = db.sqlite.prepare(
      "INSERT INTO memories (id, scope, project_id, blueprint_id, body, source, origin_id) VALUES (?, ?, ?, ?, ?, 'distilled', ?)",
    );
    for (const m of fresh) {
      enforceMemoryCap(db, m.scope === "project" ? { projectId } : { blueprintId });
      insert.run(newId("mem"), m.scope, projectId, m.scope === "blueprint" ? blueprintId : null, m.body, runId);
    }
  } catch (err) {
    console.error(`[worker] distillation failed for run ${runId}: ${err instanceof Error ? err.message : String(err)}`);
  }
}

/**
 * Cap active memories at 30 within one scope by disabling the oldest active row.
 * Project scope keys on `project_id`; blueprint scope on `blueprint_id`.
 */
function enforceMemoryCap(db: RunoffDb, scope: { projectId: string } | { blueprintId: string }): void {
  const clause =
    "projectId" in scope ? "scope='project' AND project_id = ?" : "scope='blueprint' AND blueprint_id = ?";
  const val = "projectId" in scope ? scope.projectId : scope.blueprintId;
  const { n } = db.sqlite
    .prepare(`SELECT COUNT(*) AS n FROM memories WHERE ${clause} AND status='active'`)
    .get(val) as { n: number };
  if (n >= 30) {
    db.sqlite
      .prepare(
        `UPDATE memories SET status='disabled' WHERE id = (SELECT id FROM memories WHERE ${clause} AND status='active' ORDER BY rowid LIMIT 1)`,
      )
      .run(val);
  }
}

File: backend/runoff_api/worker/run_loop.py
"""Port of apps/worker/src/runLoop.ts — the claim loop, engine IO, and
post-run distillation. TS wins on every statement.

Runtime shapes are plain camelCase-keyed dicts; execution is synchronous.
`make_engine_io` returns an object satisfying the engine's duck-typed `io`
protocol (`emit`/`poll_inputs`/`sleep`).
"""

import json
import sys
import time
from types import SimpleNamespace

from runoff_api.core.db import RunoffDb
from runoff_api.core.ids import new_id
from runoff_api.core.jsonutil import to_json
from runoff_api.core.previous_run import previous_completed_document
from runoff_api.core.types.blueprint import BlueprintContent
from runoff_api.engine.distill import distill_run
from runoff_api.engine.run import execute_run
from runoff_api.worker.resolve_sources import resolve_run_sources
from runoff_api.worker.run_data import build_run_data

# Atomically claim the oldest queued run (single statement; RETURNING gives us
# its identity). Byte-copied from CLAIM_SQL in runLoop.ts, aliases included.
CLAIM_SQL = """
UPDATE runs SET status='running', started_at=datetime('now')
WHERE id = (SELECT id FROM runs WHERE status='queued' ORDER BY created_at LIMIT 1)
RETURNING id, blueprint_id AS blueprintId, blueprint_rev AS blueprintRev, period"""


def claim_queued_run(db: RunoffDb) -> dict | None:
    row = db.execute(CLAIM_SQL).fetchone()
    return dict(row) if row is not None else None


def make_engine_io(db: RunoffDb, run_id: str):
    """The side-channel the engine drives for a single run. `emit` appends a run
    event (seq = max+1) in one transaction and mirrors run status onto the
    `runs` row (and, for flags, into the `flags` table) so the API can report
    status without replaying events. `poll_inputs` drains pending worker inputs.
    """

    def emit(e: dict) -> None:
        # Immediate mode takes the write lock up front, so the MAX(seq) read and
        # the insert see one snapshot — avoids SQLITE_BUSY_SNAPSHOT when the web
        # app writes concurrently. open_db runs in autocommit (isolation_level=None),
        # so BEGIN IMMEDIATE / COMMIT are issued explicitly to make the txn real.
        db.execute("BEGIN IMMEDIATE")
        try:
            seq = db.execute(
                "SELECT COALESCE(MAX(seq), 0) + 1 AS seq FROM run_events WHERE run_id = ?",
                (run_id,),
            ).fetchone()["seq"]
            db.execute(
                "INSERT INTO run_events (run_id, seq, type, payload) VALUES (?, ?, ?, ?)",
                (run_id, seq, e["type"], to_json(e)),
            )
            t = e["type"]
            if t == "flag_raised":
                # The engine mints per-run flag ids (`flag_1`, …); the `flags`
                # table is globally keyed, so namespace the row id by run to avoid
                # a cross-run primary-key collision. The event payload keeps the
                # bare id; the API reads the row id back.
                db.execute(
                    "INSERT INTO flags (id, run_id, code, section_key, question, options) "
                    "VALUES (?, ?, ?, ?, ?, ?)",
                    (
                        f"{run_id}_{e['flagId']}",
                        run_id,
                        e["code"],
                        e["sectionKey"],
                        e["question"],
                        to_json(e["options"]),
                    ),
                )
            elif t == "paused":
                db.execute("UPDATE runs SET status=? WHERE id = ?", ("paused", run_id))
            elif t == "resumed":
                db.execute("UPDATE runs SET status=? WHERE id = ?", ("running", run_id))
            elif t == "run_completed":
                db.execute(
                    "UPDATE runs SET status='complete', finished_at=datetime('now'), "
                    "stats=?, document=? WHERE id = ?",
                    (to_json(e["stats"]), to_json(e["document"]), run_id),
                )
            elif t == "run_failed":
                db.execute(
                    "UPDATE runs SET status='failed', finished_at=datetime('now') WHERE id = ?",
                    (run_id,),
                )
            db.execute("COMMIT")
        except Exception:
            # Guard on in_transaction: SQLite may have already auto-rolled back
            # (e.g. SQLITE_FULL), and an explicit ROLLBACK then raises "no
            # transaction is active", masking the original error. Mirrors
            # better-sqlite3's .transaction() wrapper, which rolls back only when
            # inTransaction.
            if db.in_transaction:
                db.execute("ROLLBACK")
            raise

    def poll_inputs() -> list[dict]:
        rows = db.execute(
            "SELECT id, kind, payload FROM run_inputs WHERE run_id = ? AND consumed_at IS NULL ORDER BY id",
            (run_id,),
        ).fetchall()
        out: list[dict] = []
        for r in rows:
            db.execute("UPDATE run_inputs SET consumed_at=datetime('now') WHERE id = ?", (r["id"],))
            payload = json.loads(r["payload"])
            out.append(
                {"kind": r["kind"], "text": payload.get("text"), "questionId": payload.get("questionId")}
            )
        return out

    def sleep(ms: float) -> None:
        time.sleep(ms / 1000)

    return SimpleNamespace(emit=emit, poll_inputs=poll_inputs, sleep=sleep)


def fail_stale_runs(db: RunoffDb) -> int:
    """Boot recovery: any run left `running`/`paused` by a crashed worker is
    unrecoverable (the engine's in-memory state is gone), so mark it failed and
    append a `run_failed` event. Returns how many runs were recovered.
    """
    stale = db.execute("SELECT id FROM runs WHERE status IN ('running', 'paused')").fetchall()
    for row in stale:
        make_engine_io(db, row["id"]).emit({"type": "run_failed", "error": "worker restarted mid-run"})
    return len(stale)


def process_one(db: RunoffDb, client) -> bool:
    """Claim and execute one queued run. Loads the pinned blueprint revision
    content and its bound source files, runs the engine (which persists success
    and, on failure, emits `run_failed` before throwing), and guards pre-engine
    throws (e.g. a malformed revision) by marking the run failed itself.
    Returns whether a run was claimed.
    """
    claimed = claim_queued_run(db)
    if not claimed:
        return False

    io = make_engine_io(db, claimed["id"])
    try:
        rev_row = db.execute(
            "SELECT content FROM blueprint_revisions WHERE blueprint_id = ? AND rev = ?",
            (claimed["blueprintId"], claimed["blueprintRev"]),
        ).fetchone()
        if rev_row is None:
            raise RuntimeError(
                f"blueprint revision not found: {claimed['blueprintId']}@{claimed['blueprintRev']}"
            )
        content = BlueprintContent.model_validate(json.loads(rev_row["content"])).model_dump(by_alias=True)

        resolved = resolve_run_sources(db, claimed["blueprintId"], claimed["period"])
        files = resolved["files"]
        gaps = resolved["gaps"]

        project_id = db.execute(
            "SELECT project_id AS projectId FROM blueprints WHERE id = ?", (claimed["blueprintId"],)
        ).fetchone()["projectId"]
        # Deduped, first-seen order (mirrors TS `[...new Set(...)]`).
        bound_family_ids = list(dict.fromkeys(fid for s in content["sections"] for fid in s["familyIds"]))
        data = build_run_data(db, project_id, bound_family_ids, claimed["period"])

        run_row = db.execute(
            "SELECT created_at AS createdAt FROM runs WHERE id = ?", (claimed["id"],)
        ).fetchone()
        previous = previous_completed_document(
            db, claimed["blueprintId"], claimed["id"], run_row["createdAt"]
        )

        memory_rows = db.execute(
            "SELECT id, body, scope FROM memories "
            "WHERE status='active' AND (blueprint_id = ? OR (scope='project' AND project_id = "
            "(SELECT project_id FROM blueprints WHERE id = ?))) "
            "ORDER BY rowid",
            (claimed["blueprintId"], claimed["blueprintId"]),
        ).fetchall()
        memories = [{"id": m["id"], "body": m["body"], "scope": m["scope"]} for m in memory_rows]

        execute_run(
            client=client,
            content=content,
            files=files,
            data=data,
            io=io,
            blueprint_rev=claimed["blueprintRev"],
            previous_document=previous["document"] if previous else None,
            memories=memories,
            period=claimed["period"],
            gaps=gaps,
        )
        # Success is already persisted by the `run_completed` emit handler.
        distill_completed_run(db, client, claimed["id"], claimed["blueprintId"], content)
    except Exception as err:  # noqa: BLE001 — mirrors the TS catch-all
        message = str(err)
        print(f"[worker] run {claimed['id']} failed: {message}", file=sys.stderr)
        # If the engine threw it already emitted `run_failed` (which set status).
        # For pre-engine throws (revision missing / validation) nothing marked it yet.
        row = db.execute("SELECT status FROM runs WHERE id = ?", (claimed["id"],)).fetchone()
        if row is not None and row["status"] != "failed":
            io.emit({"type": "run_failed", "error": message})
    return True


def distill_completed_run(db: RunoffDb, client, run_id: str, blueprint_id: str, content: dict) -> None:
    """Post-run learning: turn this run's human interventions into standing
    memories. Read-side is the run's own event log + resolved flags; any failure
    is logged and swallowed — distillation must never affect run status.
    """
    try:
        events = db.execute(
            "SELECT type, payload FROM run_events WHERE run_id = ? AND "
            "type IN ('steer_received','question_raised','question_answered') ORDER BY seq",
            (run_id,),
        ).fetchall()
        # `question_answered` carries only {questionId, answer}; the question text
        # lives in the preceding `question_raised` event, so map ids -> text first.
        question_text: dict[str, str] = {}
        for e in events:
            if e["type"] != "question_raised":
                continue
            p = json.loads(e["payload"])
            if isinstance(p.get("questionId"), str) and isinstance(p.get("question"), str):
                question_text[p["questionId"]] = p["question"]
        interactions: dict = {"steers": [], "answers": [], "flagResolutions": []}
        for e in events:
            p = json.loads(e["payload"])
            if e["type"] == "steer_received" and isinstance(p.get("text"), str):
                interactions["steers"].append(p["text"])
            if e["type"] == "question_answered" and isinstance(p.get("answer"), str):
                qid = p.get("questionId")
                interactions["answers"].append(
                    {
                        "question": question_text.get(qid, str(qid if qid is not None else "")),
                        "answer": p["answer"],
                    }
                )
        flag_rows = db.execute(
            "SELECT question, resolution FROM flags WHERE run_id = ? AND status = 'resolved' "
            "AND resolution IS NOT NULL",
            (run_id,),
        ).fetchall()
        interactions["flagResolutions"] = [
            {"question": f["question"], "resolution": f["resolution"]} for f in flag_rows
        ]

        if not interactions["steers"] and not interactions["answers"] and not interactions["flagResolutions"]:
            return

        # This blueprint's project, so project-scoped memories land on the right row.
        project_row = db.execute(
            "SELECT project_id AS projectId FROM blueprints WHERE id = ?", (blueprint_id,)
        ).fetchone()
        project_id = (
            project_row["projectId"]
            if project_row is not None and project_row["projectId"] is not None
            else ""
        )

        # Dedup pool spans both scopes (dedup is on lowercased body, scope-agnostic).
        existing = db.execute(
            "SELECT body, scope FROM memories WHERE blueprint_id = ? OR (scope='project' AND project_id = ?)",
            (blueprint_id, project_id),
        ).fetchall()
        existing_list = [{"body": r["body"], "scope": r["scope"]} for r in existing]

        fresh = distill_run(
            client=client,
            title=content["title"],
            section_headings=[s["heading"] for s in content["sections"]],
            interactions=interactions,
            existing=existing_list,
        )

        insert_sql = (
            "INSERT INTO memories (id, scope, project_id, blueprint_id, body, source, origin_id) "
            "VALUES (?, ?, ?, ?, ?, 'distilled', ?)"
        )
        for m in fresh:
            enforce_memory_cap(
                db, {"projectId": project_id} if m["scope"] == "project" else {"blueprintId": blueprint_id}
            )
            db.execute(
                insert_sql,
                (
                    new_id("mem"),
                    m["scope"],
                    project_id,
                    blueprint_id if m["scope"] == "blueprint" else None,
                    m["body"],
                    run_id,
                ),
            )
    except Exception as err:  # noqa: BLE001 — distillation must never affect run status
        print(f"[worker] distillation failed for run {run_id}: {err}", file=sys.stderr)


def enforce_memory_cap(db: RunoffDb, scope: dict) -> None:
    """Cap active memories at 30 within one scope by disabling the oldest active
    row. Project scope keys on `project_id`; blueprint scope on `blueprint_id`.
    """
    if "projectId" in scope:
        clause = "scope='project' AND project_id = ?"
        val = scope["projectId"]
    else:
        clause = "scope='blueprint' AND blueprint_id = ?"
        val = scope["blueprintId"]
    n = db.execute(
        f"SELECT COUNT(*) AS n FROM memories WHERE {clause} AND status='active'", (val,)
    ).fetchone()["n"]
    if n >= 30:
        db.execute(
            f"UPDATE memories SET status='disabled' WHERE id = "
            f"(SELECT id FROM memories WHERE {clause} AND status='active' ORDER BY rowid LIMIT 1)",
            (val,),
        )
